File: packages/tracelet_ios/ios/Classes/TraceletLogger.swift
import Foundation

/// Dual (console + SQLite) logger for the Tracelet plugin.
///
/// Respects configured log level. Provides getLog, pruneOldLogs, and email export.
final class TraceletLogger {
    private let configManager: ConfigManager
    private let database: TraceletDatabase

    /// Log levels: OFF(0), ERROR(1), WARNING(2), INFO(3), DEBUG(4), VERBOSE(5)
    enum Level: Int, Comparable {
        case off = 0, error = 1, warning = 2, info = 3, debug = 4, verbose = 5

        var label: String {
            switch self {
            case .off: return "OFF"
            case .error: return "ERROR"
            case .warning: return "WARNING"
            case .info: return "INFO"
            case .debug: return "DEBUG"
            case .verbose: return "VERBOSE"
            }
        }

        static func from(_ string: String) -> Level {
            switch string.uppercased() {
            case "ERROR": return .error
            case "WARNING", "WARN": return .warning
            case "INFO": return .info
            case "DEBUG": return .debug
            case "VERBOSE": return .verbose
            default: return .info
            }
        }

        static func < (lhs: Level, rhs: Level) -> Bool {
            return lhs.rawValue < rhs.rawValue
        }
    }

    init(configManager: ConfigManager, database: TraceletDatabase) {
        self.configManager = configManager
        self.database = database
    }

    // MARK: - Logging methods

    func error(_ message: String) {
        log(.error, message)
    }

    func warning(_ message: String) {
        log(.warning, message)
    }

    func info(_ message: String) {
        log(.info, message)
    }

    func debug(_ message: String) {
        log(.debug, message)
    }

    func verbose(_ message: String) {
        log(.verbose, message)
    }

    /// Log from Dart side with string level.
    func log(levelString: String, message: String) {
        let level = Level.from(levelString)
        log(level, message, source: "dart")
    }

    // MARK: - Query

    func getLog(query: [String: Any]? = nil) -> [[String: Any]] {
        return database.getLogs(query: query)
    }

    func getLogForEmail() -> String {
        return database.getLogForEmail()
    }

    func destroyLog() -> Bool {
        return database.deleteAllLogs()
    }

    func pruneOldLogs() {
        let maxDays = configManager.getLogMaxDays()
        database.pruneOldLogs(maxDays: maxDays)
    }

    // MARK: - Core

    private func log(_ level: Level, _ message: String, source: String = "plugin") {
        let configLevel = Level(rawValue: configManager.getLogLevel()) ?? .verbose

        // Skip if level is above configured threshold
        guard level.rawValue <= configLevel.rawValue, level != .off else { return }

        // Console log
        NSLog("[Tracelet] [\(level.label)] \(message)")

        // SQLite log
        database.insertLog(level: level.label, message: message, source: source)
    }
}

File: sdk/android/tracelet-sdk/src/main/kotlin/com/ikolvi/tracelet/sdk/util/TraceletLogger.kt
package com.ikolvi.tracelet.sdk.util

import android.content.Context
import android.util.Log
import com.ikolvi.tracelet.sdk.ConfigManager

/**
 * Logger that writes to Android logcat.
 *
 * Log levels: OFF(0), ERROR(1), WARNING(2), INFO(3), DEBUG(4), VERBOSE(5)
 */
class TraceletLogger(
    private val context: Context,
    private val config: ConfigManager,
) {
    var rustDatabase: uniffi.tracelet_core.DatabaseManager? = null

    companion object {
        private const val TAG = "Tracelet"

        // Log level constants matching Dart LogLevel enum
        const val LEVEL_OFF = 0
        const val LEVEL_ERROR = 1
        const val LEVEL_WARNING = 2
        const val LEVEL_INFO = 3
        const val LEVEL_DEBUG = 4
        const val LEVEL_VERBOSE = 5

        fun levelToString(level: Int): String = when (level) {
            LEVEL_ERROR -> "ERROR"
            LEVEL_WARNING -> "WARNING"
            LEVEL_INFO -> "INFO"
            LEVEL_DEBUG -> "DEBUG"
            LEVEL_VERBOSE -> "VERBOSE"
            else -> "OFF"
        }
    }

    /** Log an error message. */
    fun error(message: String, tag: String = TAG) {
        logInternal(LEVEL_ERROR, message, tag)
    }

    /** Log a warning message. */
    fun warning(message: String, tag: String = TAG) {
        logInternal(LEVEL_WARNING, message, tag)
    }

    /** Log an info message. */
    fun info(message: String, tag: String = TAG) {
        logInternal(LEVEL_INFO, message, tag)
    }

    /** Log a debug message. */
    fun debug(message: String, tag: String = TAG) {
        logInternal(LEVEL_DEBUG, message, tag)
    }

    /** Log a verbose message. */
    fun verbose(message: String, tag: String = TAG) {
        logInternal(LEVEL_VERBOSE, message, tag)
    }

    /** Log with explicit level (from Dart). */
    fun log(level: String, message: String): Boolean {
        // Compare without allocating a new uppercase string (A-L2).
        val levelInt = when {
            level.equals("ERROR", ignoreCase = true) -> LEVEL_ERROR
            level.equals("WARNING", ignoreCase = true) || level.equals("WARN", ignoreCase = true) -> LEVEL_WARNING
            level.equals("INFO", ignoreCase = true) -> LEVEL_INFO
            level.equals("DEBUG", ignoreCase = true) -> LEVEL_DEBUG
            level.equals("VERBOSE", ignoreCase = true) -> LEVEL_VERBOSE
            else -> LEVEL_INFO
        }
        logInternal(levelInt, message, TAG)
        return true
    }

    /** Get the log as a string (optionally filtered). */
    fun getLog(query: Map<String, Any?>?): String {
        return try {
            val limit = (query?.get("limit") as? Number)?.toInt() ?: 500
            val logs = rustDatabase?.getLogs(limit) ?: emptyList()
            logs.joinToString("\n") { "[${it.timestamp}] [${it.level}] ${it.message}" }
        } catch (e: Exception) {
            "Failed to retrieve logs: ${e.message}"
        }
    }

    /** Delete all log entries. */
    fun destroyLog(): Boolean {
        return try {
            rustDatabase?.clearLogs()
            true
        } catch (e: Exception) {
            false
        }
    }

    /** Email the log (returns intent data; actual email launch is done by caller). */
    fun getLogForEmail(): String {
        return getLog(mapOf("limit" to 2000))
    }

    /** Prune old logs based on config. */
    fun pruneOldLogs() {
        // Not implemented
    }

    // =========================================================================
    // Private
    // =========================================================================

    private fun logInternal(level: Int, message: String, tag: String) {
        val configLevel = config.getLogLevel()
        if (configLevel == LEVEL_OFF || level > configLevel) return

        // Write to Android logcat
        when (level) {
            LEVEL_ERROR -> Log.e(tag, message)
            LEVEL_WARNING -> Log.w(tag, message)
            LEVEL_INFO -> Log.i(tag, message)
            LEVEL_DEBUG -> Log.d(tag, message)
            LEVEL_VERBOSE -> Log.v(tag, message)
        }

        // Persist to Rust Core Database
        try {
            val levelName = levelToString(level)
            rustDatabase?.insertLog(levelName, message, "plugin")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to persist log to Rust Database: ${e.message}")
        }
    }
}

